File: packages/maplibre_flutter_android/android/src/main/java/dev/maplibreflutter/maplibre_flutter_android/MapRegistry.java
package dev.maplibreflutter.maplibre_flutter_android;

import androidx.annotation.Keep;
import java.util.HashMap;
import java.util.Map;

/**
 * Maps the Dart-generated {@code mapId} to its {@link MapLibreController}.
 *
 * <p>The Dart side mints a {@code mapId}, passes it in the {@code AndroidView}
 * creation params, and later looks the controller up here over jnigen — so
 * control flows Dart → jni → controller with no data-path method channel
 * (CLAUDE.md §10) and without coordinating the engine's platform-view id.
 *
 * <p>Java, not Kotlin — see {@link MapLibreController} for why.
 */
@Keep
public final class MapRegistry {
  private MapRegistry() {}

  private static final Map<Integer, MapLibreController> CONTROLLERS = new HashMap<>();

  /** Create and register a controller for {@code mapId} (called from the view factory). */
  @Keep
  public static synchronized MapLibreController register(int mapId) {
    MapLibreController controller = new MapLibreController();
    CONTROLLERS.put(mapId, controller);
    return controller;
  }

  /** The controller for {@code mapId}, or null if the view does not exist (yet). */
  @Keep
  public static synchronized MapLibreController get(int mapId) {
    return CONTROLLERS.get(mapId);
  }

  @Keep
  public static synchronized void unregister(int mapId) {
    CONTROLLERS.remove(mapId);
  }
}

File: packages/maplibre_flutter_ios_sdk/ios/maplibre_flutter_ios_sdk/Sources/maplibre_flutter_ios_sdk/MapRegistry.swift
import Foundation

/// Maps the Dart-minted `mapId` to its [MapLibreController].
///
/// Dart mints a `mapId`, passes it via `creationParams`, and looks the
/// controller up here over swiftgen — control flows Dart → ObjC runtime →
/// controller with no data-path method channel (CLAUDE.md §10). Foundation-only
/// so swiftgen can compile it (see [MapLibreController]).
// Plain @objc: runtime name is module-qualified to match the generated bindings
// (see MapLibreController for why).
@objc public class MapRegistry: NSObject {
  private static var controllers: [Int: MapLibreController] = [:]
  private static let lock = NSLock()

  /// Create and register a controller for `mapId` (called from the view factory).
  @objc public static func register(_ mapId: Int) -> MapLibreController {
    let controller = MapLibreController()
    lock.lock()
    controllers[mapId] = controller
    lock.unlock()
    return controller
  }

  /// The controller for `mapId`, or nil if the view does not exist (yet).
  @objc public static func get(_ mapId: Int) -> MapLibreController? {
    lock.lock()
    defer { lock.unlock() }
    return controllers[mapId]
  }

  @objc public static func unregister(_ mapId: Int) {
    lock.lock()
    controllers.removeValue(forKey: mapId)
    lock.unlock()
  }
}
